tools/castle-editor/code/framedesign.pas

```diff
@@ -1495,9 +1574,37 @@ procedure TDesignFrame.PropertyGridModified(Sender: TObject);
     end;
   end;
 
+  if not UndoSystem.ScheduleRecordUndoOnRelease then
+  begin
+    { here we have PropertyGridModified conflict with Schedule Record Undo OnRelease
+      So we should ignore changes to PropertyGrid in case the change is caused
+      by dragging, otherwise we'll record an undo for every OnMotion of dragging }
+    if Sender is TOICustomPropertyGrid then
+      RecordUndo('Change ' + TOICustomPropertyGrid(Sender).GetActiveRow.Name + ' to ' + TOICustomPropertyGrid(Sender).CurrentEditValue, TOICustomPropertyGrid(Sender).ItemIndex)
+    else
+      RecordUndo('');
+  end;
+
   MarkModified;
 end;
 
+procedure TDesignFrame.RecordUndo(const UndoComment: String; const ItemIndex: Integer = -1);
+var
+  T: TDateTime;
+  SelectedName: String;
+  SelectedC: TComponent;
+begin
+  T := Now;
+  SelectedC := GetSelectedComponent;
+  if (SelectedC <> nil) then
+    SelectedName := SelectedC.Name
+  else
+    SelectedName := '';
+
+  UndoSystem.RecordUndo(ComponentToString(FDesignRoot), SelectedName, ItemIndex, ControlProperties.TabIndex, UndoComment);
+  WriteLnLog('Undo recorded in ' + FloatToStr((Now - T) * 24 * 60 * 60) + 's for ' + SelectedName);
```

Comment:
Use instead `Timer` to measure time of the operation, https://castle-engine.io/apidoc-unstable/html/CastleTimeUtils.html#Timer . Should look like this:


```pascal
var
  TimeStart: TTimerResult;
  SelectedName: String;
  SelectedC: TComponent;
begin
  TimeStart := Timer;

  SelectedC := GetSelectedComponent;
  if (SelectedC <> nil) then
    SelectedName := SelectedC.Name
  else
    SelectedName := '';

  UndoSystem.RecordUndo(ComponentToString(FDesignRoot), SelectedName, ItemIndex, ControlProperties.TabIndex, UndoComment);

  WritelnLog('Undo recorded in %fs for %s', [TimeStart.ElapsedTime, SelectedName]);
end;
```

We use `Timer` across CGE for such measurements.

Reply:
Done!